features/lockscreen/impl/src/main/kotlin/io/element/android/features/lockscreen/impl/biometric/BiometricUnlock.kt

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2023 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.element.android.features.lockscreen.impl.biometric
+
+import android.security.keystore.KeyPermanentlyInvalidatedException
+import androidx.biometric.BiometricPrompt
+import androidx.biometric.BiometricPrompt.CryptoObject
+import androidx.biometric.BiometricPrompt.PromptInfo
+import androidx.core.content.ContextCompat
+import androidx.fragment.app.FragmentActivity
+import io.element.android.libraries.cryptography.api.EncryptionDecryptionService
+import io.element.android.libraries.cryptography.api.SecretKeyRepository
+import kotlinx.coroutines.CompletableDeferred
+import javax.crypto.Cipher
+
+interface BiometricUnlock {
+
+    interface Callback {
+        fun onBiometricSetupError()
+        fun onBiometricUnlockSuccess()
+        fun onBiometricUnlockFailed(error: Exception?)
+    }
+
+    sealed interface AuthenticationResult {
+        data object Success : AuthenticationResult
+        data class Failure(val error: Exception? = null) : AuthenticationResult
+    }
+
+    val isActive: Boolean
+    fun setup()
+    suspend fun authenticate(): AuthenticationResult
+}
+
+class NoopBiometricUnlock : BiometricUnlock {
+    override val isActive: Boolean = false
+    override fun setup() = Unit
+    override suspend fun authenticate() = BiometricUnlock.AuthenticationResult.Failure()
+}
+
+class DefaultBiometricUnlock(
+    private val activity: FragmentActivity,
+    private val promptInfo: PromptInfo,
+    private val secretKeyRepository: SecretKeyRepository,
+    private val encryptionDecryptionService: EncryptionDecryptionService,
+    private val keyAlias: String,
+    private val callbacks: List<BiometricUnlock.Callback>
+) : BiometricUnlock {
+
+    override val isActive: Boolean = true
+
+    private lateinit var cryptoObject: CryptoObject
+
+    override fun setup() {
+        try {
+            val secretKey = ensureKey()
+            val cipher = encryptionDecryptionService.createEncryptionCipher(secretKey)
```

Comment:
While this will work in general, on EA [we had some nasty issues](https://github.com/vector-im/element-android/issues/7064#issuecomment-1241629631) with 'user not authenticated' errors. To avoid this, one possible workaround may be to create the `Cipher` on demand and have 2 internal authentication methods, since it seems like this is a possible workaround:

1. `authenticateWithCryptoObject`, should be called by default, it will do what's inside `authenticate` right now. If the `createEncryptionCipher` inside this authenticate method fails with the 'user not authenticated error', however, we go to the 2nd method.
2. `authenticateWithoutCryptoObject`, will call the same `prompt.authenticate` method without a `CryptoObject`, which will authenticate the user and get rid of the error, then you should be able to initialize the `Cipher` and use it as you'd usually do.

I'm not saying you should implement this now, but we definitely should if we see those kind of issues coming back on EXA.

Reply:
`user not authenticated error` is something different than `KeyPermanentlyInvalidatedException`?

Reply:
Ok I see, otherwise I'll remove the flag for now...

Reply:
If you do that we might have to handle key migrations in the future, so be careful.

Reply:
Will handle that later if we encounter the issue, no time for now :/

Reply:
No worries, given the deadlines, I think that's the right call too